### ISSUE TITLE
Restoring Transcoder State

### ISSUE THREAD

**Issue by j0sh** [livepeer/go-livepeer#353]
General discussion thread for saving and restoring transcoder state. Some of this may apply to broadcasters, but transcoders are the focus for now.

Record pertinent bits of information as jobs come in. TxID, JobID, ManifestID, BroadcasterID, TranscoderID, StartBlock, EndBlock, ClaimedAt, VerifiedAt, ...?

Naturally, the `TranscoderID` would be redundant for a given transcoder's record-keeping, but could be useful if we share the same table for broadcasters, and avoids a small ETL step if someone wants to merge multiple transcoder tables together later.

Two distinct stages:

* Re-subscribing to a job that might be in progress. We already have a record of the job.

*  Subscribing to jobs that were 'missed' while the node was down. Involves replaying the eth event log.

Claims and Rewards. Anything particular we need to consider here?

### Implementation
* SQLite
* Allows nodes a local record of their job history, claims, errors, etc.
* Separate table to record transcoding errors or exceptions as they happen, for later review?
* Can migrate the peer cache (currently a flat file) to SQLite as well.
  * Could also note the type of peer? Transcoder, broadcaster, boot/relay?
* Can also have a general "KV" table with one-off data
  * The last block seen in order to replay the Eth log?
  * Last round that fees/rewards were claimed?
  * Anything else?

**Comment by jozanza** [comment]
I was actually thinking it might make sense to keep it simple and read the data from the chain via rpc provider since all the info we'd store is already written there.

**Comment by dob** [comment]
Great proposal @j0sh. Is there a chance that transcoded segments that haven't yet been claimed can be reclaimed as well upon restart when restoring from this state? For example if I see that I was assigned to job X, and I don't have a lastClaim segment, but I see that I've encoded and temporarily stored segments 0-90, it would be great to claim them at this time.

The other thing I can think of is that we slash if you miss a required call to verify(), so perhaps this needs to be reconciled with the state as well. It may be rare because the node will usually call verify very shortly after making a claim, but if for some reason it hasn't, it could potentially tell this from state and make the call to avoid being slashed (only if the verificaiton period is still open).

**Comment by j0sh** [comment]
>  read the data from the chain via rpc provider since all the info we'd store is already written there.

@jozanza  Perhaps, but wouldn't we need to re-scan the chain for an indefinite number of blocks in order to re-discover long-running jobs?

>  Is there a chance that transcoded segments that haven't yet been claimed can be reclaimed as well upon restart when restoring from this state

@dob Yep, that should be one of the goals.

There's a couple ways to implement this. One is attaching those fields to the job itself, where we have `lastSegmentTranscoded` / `lastClaim` / `lastVerified` . If we want to keep track of segments on an individual basis in a separate table, we could blockstamp/timestamp each claim, eg `claimAt` (and `verifiedAt` when appropriate). Tracking individual segments is probably a bit more work, but might make things like error accounting a bit more consistent, and we can include more 'interesting' data for historical/analytical purposes, eg how long it took to transcode a given segment, the # of frames, etc.

**Comment by yondonfu** [comment]
> Perhaps, but wouldn't we need to re-scan the chain for an indefinite number of blocks in order to re-discover long-running jobs?

Related: https://github.com/livepeer/go-livepeer/issues/315 The scheme described in that issue would handle closed websocket connections that break subscriptions so the node can just keep track of the block at which a subscription ended and then filter logs starting from that block when it re-subscribes to make sure it discovers events that were relevant to it that were emitted in the period of time that it was not subscribed to the event. However, the scheme wouldn't work if the node crashes. We could modify it to keep track of the block at which each event was received. So for jobs, there could be a `lastReceivedNewJob` field stored locally and if the transcoder crashes or its subscription breaks for some reason, when it re-subscribes it filters for logs starting from the `lastReceivedNewJob` block.

> Is there a chance that transcoded segments that haven't yet been claimed can be reclaimed as well upon restart when restoring from this state? 

We will need the [transcode receipt](https://github.com/livepeer/wiki/blob/master/SPEC.md#transcode-receipt) for each of those segments. Since we would need to keep around the input segment data for any segment that is included as a part of a claim just in case that segment is challenged for verification, we could not store `dataHash`, but instead just generate `keccak256(inputSegmentData)` on the fly when creating the Merkle root for a claim. At the moment since bitexact verification only needs the input segment data, we don't need to keep around the actual transcoded segment data, since verification only cares about `keccak256(transcodedSegmentData)`, so the node could just store `transcodedDataHash` and discard the transcoded segment data. However, in the future as we explore different verification approaches, the node might need to keep around the actual transcoded segment data (ex. if instead of transcoding the input segment data, the verifier decodes the transcoded segment data and runs some comparison checks with the input segment data. In this case, the node needs to hold onto both the input segment data and transcoded segment data in case a particular segment is challenged for verification).

If a transcoder goes offline and reboots while retaining the necessary data for submitting claims it could first query the JobsManager contract to check if it has submitted a claim for a job before. If it has not submitted a claim for the job before, it needs to make sure that it has not been more than 256 blocks since the job's creation. If it has been more than 256 blocks since the job's creation, the transcoder cannot submit the claim.

On a related note, the node's claim manger currently holds onto input segment data in memory - if we are keeping around input segment data for future recoveries, perhaps it just reads from storage

> The other thing I can think of is that we slash if you miss a required call to verify(), so perhaps this needs to be reconciled with the state as well.

If you know the `claimID` for your recent claims and their associated segment data, you could iterate through each segment number for that claim and check if it was challenged for verification

**Comment by jozanza** [comment]
Unless there is a critical bug that would immediately get fixed by this enhancement, perhaps we should hold off discussing implementation details until @f1l1b0x shares his workflow proposal.

**Comment by ericxtang** [comment]
The local storage project can be a pretty big effort, so I'd like to propose to include the most immediate fixes in this issue, and take an iterative approach for the long term.

The problems we currently have:
* If the transcoder crashes or gets restarted in the middle of a transcoding job, there is no way for it to recover the states of that job.  This can potentially get the transcoder slashed (if it has done the claim but not the verification yet)
* We sometimes miss events because of the websocket instability
* We sometimes miss events if the transcoder node crashes or we need to take it offline briefly for maintenance

The first problem seems most crucial - since we don't want the transcoder to be slashed due to a node crash.  But maybe it doesn't require lot of incremental effort to address the other problems.

I see this project as a bug fix before the network launch, and @f1l1b0x's proposal as a phase 2 effort that will require some more planning.

**Comment by j0sh** [comment]
> Unless there is a critical bug that would immediately get fixed by this enhancement, perhaps we should hold off discussing implementation details until @f1l1b0x shares his workflow proposal.

https://github.com/livepeer/go-livepeer/issues/269 is marked Bug / Stability / Network Launch . I was hoping to close it by the mainnet release. Even if we have to revisit all this after the launch, having a more resilient node at the outset will buy us some more time to hone in on the next design. Still, happy to table this entire discussion if others feel like it's not a launch priority.

> node's claim manger currently holds onto input segment data in memory - if we are keeping around input segment data for future recoveries, perhaps it just reads from storage

Good point, we absolutely should read from disk in that case.

> So for jobs, there could be a lastReceivedNewJob field stored locally and if the transcoder crashes or its subscription breaks for some reason, when it re-subscribes it filters for logs starting from the lastReceivedNewJob block.

Yes, storing `lastReceivedNewJob` (or equivalently, `lastBlockProcessed`) would be useful for picking up jobs that were missed while the node was down. Good proposal in https://github.com/livepeer/go-livepeer/issues/315.

> We will need the transcode receipt for each of those segments.

Good to know, so we'll need to store those on disk as well.

**Comment by j0sh** [comment]
> The local storage project can be a pretty big effort ... The first problem seems most crucial - since we don't want the transcoder to be slashed due to a node crash.

In order to recover from that we need to persist information on disk (job info, transcode receipts).

The last two items for "missing events" can be done by replaying the eth log; it doesn't need local storage AFAICS.

**Comment by j0sh** [comment]
Does anyone have strong feelings one way or the other regarding DB normalization of claims vs segments / receipts? Here are some options:

* `segments` table with `claimedAt`  (submitted) / `claimTxAt` (observed on the blockchain) columns for each segment
* Separate `claims` table that incorporates the segment range for each claim. This more closely matches the domain model but is perhaps a bit less intuitive to query, and probably leaves more room for inconsistencies at the DB level unless we're meticulous about constraints (eg, preventing overlapping ranges)
* Or even a `claims` table with a row for each segment, but in this case I'd prefer to sacrifice a bit of relational purity and instead have nullable `claimedAt` / `claimTxAt` fields in the `segments` table.

**Comment by ericxtang** [comment]
I think we should have both `segments` and `claims`.  We might not have to store all the segments, but I think we would need at least all the transcoded segments due to the need to have transcoded segment hash for verification.  `claims` can be used in case the transaction fails for some reason and we have to re-submit the claim.

**Comment by j0sh** [comment]
Okay, we can try having a separate `claims` table.

The question becomes: should we store data for each transcoded segment (eg, one row per profile, per segment), or something more similar to the transcode receipt that just has the concatenated hash of all transcoded profiles? 

With storing each transcoded segment, we'd get the opportunity for much more granular reporting, but would be looking at some redundant data, eg storing the broadcaster signature and source-data hash in each row [1] (unless we normalize that even further).

The major issue with storing receipts right now is that the concatenated transcoded data hashes are not actually created until claim time, so if we were to create the receipt data hashes immediately post-transcode, we'd need to rearrange some plumbing. The more I consider it, that might be the best way to do things.

[1] For the future, I think we can drop the broadcaster sig and source data hash from the claim entirely, since the verification process should still be checking those independently when retrieving the source data. Less on-chain data storage required during verification, and simplifies the issue with this schema.

**Comment by j0sh** [comment]
> if we were to create the receipt data hashes immediately post-transcode, we'd need to rearrange some plumbing

Here is roughly what these plumbing changes will end up looking like. https://github.com/livepeer/go-livepeer/pull/358/commits/71343554a7866d87c7116b153e19796e721805a3 

From the commit log:

> Call AddReceipt once per source segment instead of incrementally
per transcoded segment, and pre-compute validity checks (eg,
profile matches) and the concatenated hashes, so we don't have
to do the computation at claim time.

@ericxtang @yondonfu  can you take a quick look and let me know if the approach is sane, and whether there may be unintended semantic changes?

Here is one semantic change I can think of: we no longer count costs on skipped segments (eg, if the segment has a mismatched set of profiles). However, the original cost calculation may actually have been a slight over-estimation, since I don't believe we collect fees for unclaimed segments. This seems to be mostly for the deposit estimate calculation, so I don't think it's a huge deal either way.

Also here's an accompanying commit adding the receipts table if it helps: https://github.com/livepeer/go-livepeer/pull/358/commits/d7685e44153de37f245c7d2943125335f10e250f . Again, WIP, untested, etc, just to give an idea.

**Comment by ericxtang** [comment]
@j0sh I think the change makes sense.  Not counting cost for skipped segments is nice.  

I'm curious about the approach to pass in the `dbInsert` call into the claimManager, vs. just inserting the record in the caller if no error was returned for `cm.AddReceipt`.  Not a big deal either way, just pointing out an opportunity to reduce complexity.

**Comment by j0sh** [comment]
Yeah, when / how to insert into the DB is a bit weird right now. We have a few choices:

* Return an assortment of values from `cm.AddReceipt` so the caller has enough information to populate the DB. `jobID`, `bHash` , `tHash`, etc. A variation is: return an incomplete `DBReceipt` struct, rather than several values.
* Pass in a few more parameters (`transcodeStart`, `transcodeEnd`, the DB handle somewhere) and do the insertion within `cm.AddReceipt` itself. This might keep the claim-related logic bit more self-contained within ClaimManager, if we can also do the recovery from within ClaimManager itself (rather than invoking the DB separately from LivepeerNode and reassembling the claims from there).
* The current insertion method via callbacks, which is perhaps an unhappy mix of the above two.

I'll keep working on the claim recovery logic, and see if the path forward becomes clearer.

**Comment by j0sh** [comment]
Regarding how to manage the DB around ClaimManager, specifically whether we should invoke the DB externally to the manager, or within it.

Creating / storing a claim involves these steps:
* Create possibly multiple ranges for a claim
* Wait for the claim block(s) to be mined (need to store / check txn state locally for this as well)
* Verify and wait if needed (need sto store / check local txn state)
* Distribute fees (not sure yet if this can be a fire-and-forget operation, but probably not)

Given how involved the claiming process is, I'm thinking that we should do as much of the DB-related work within the claimmanager itself, including receipt management. That corresponds to this option

> "Pass in a few more parameters (transcodeStart, transcodeEnd, the DB handle somewhere) and do the insertion within cm.AddReceipt itself."

What do you think @ericxtang @yondonfu ?

**Comment by yondonfu** [comment]
> Distribute fees (not sure yet if this can be a fire-and-forget operation, but probably not)

Once the claim is submitted on-chain, the transcoder should wait through `verificationPeriod + slashingPeriod` blocks and only after those number of blocks will it be able to distribute fees associated with the particular claim. So, the transcoder would have to monitor the number of blocks that elapsed since a claim was made to determine when to distribute fees, but the distribute fees action itself can be fire and forget.

> Given how involved the claiming process is, I'm thinking that we should do as much of the DB-related work within the claimmanager itself, including receipt management.

I think this makes sense!

**Comment by j0sh** [comment]
Awesome, good to know. Is there any possibility that the 'distribute fee' transaction may get orphaned or stuck in mempool purgatory (maybe due to insufficient gas, uncles, etc), and should we have the node retry the transaction? This is a concern with most of the transactions we're doing, not just distribute fees.

**Comment by yondonfu** [comment]
Hm so I think there are 3 ways a transaction can "fail": 1) it can fail upon submission - this is usually due to a misconfigured transaction i.e. incorrect nonce, insufficient gas provided resulting in an always failing transaction due to out of gas or an issue with the RPC client (i.e. Infura not available) 2) it can actually fail due to a revert in the EVM - we would know this after the transaction has been mined (we do this check in `CheckTx`)` 3) it can hang around in the tx mempool after submission likely because the gas price provided was not high enough for a miner to take it.

I think there's also another scenario where there is a reorg and the state transition triggered by a successful transaction is no longer recorded on the main chain...

I think for 2) there is usually going to be something wrong about how the transaction is being crafted so we probably shouldn't retry. For 1) if the failed submission is due to an unavailable RPC client it could make sense to retry, but if it is due to the transaction being crafted incorrectly it might not make sense to retry. For 3), it probably makes sense to retry but we would need to make sure to up the gas price to make sure to invalidate the old transaction that is using the same nonce.

Perhaps we can continue the discussion on retrying for transactions in a separate issue?

**Comment by j0sh** [comment]
Merged in https://github.com/livepeer/go-livepeer/pull/358 . Thanks everybody for the discussion and review!
